Made RqMultipart.Base delete its temporary files as soon as possible.

Bertram25 opened the pull request:
For #254

I thus added a decorator to InputStream to permit deletion of the temp file
when calling close() as advised by @dmzaytsev

Thanks to @yegor256 and @dmzaytsev for the hints.

I also added a test to the new `TempInputStream` class to actually test the change.

davvd commented:
@bertram25 Thanks, let me find someone who can review this pull request

davvd commented:
@pinaf it's yours, please review

pinaf commented on src/main/java/org/takes/rq/TempInputStream.java:
```diff
@@ -0,0 +1,98 @@
+ *
+ * @author Yegor Bugayenko (yegor@teamed.io)
+ * @version $Id$
+ * @since 0.16
```
@Bertram25 0.21 here

pinaf commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -368,4 +369,31 @@ public void handlesRequestInTime() throws IOException {
+     * @throws IOException if some problem occurs.
+     */
+    @Test
+    public void testTempFileDeletion() throws IOException {
```
@Bertram25 `deletesTempFile()`

pinaf commented on src/test/java/org/takes/rq/RqMultipartTest.java:
```diff
@@ -368,4 +369,31 @@ public void handlesRequestInTime() throws IOException {
     }
 
+    /**
+     * Test the TempInputStream class for correct deletion of the temp file,
```
@Bertram25 `RqMultiPart can delete the underlying temporary file`

pinaf commented:
@Bertram25 3 comments above. Also, the test you added does not test `RqMultiPart` but `TempInputStream`. We should do the following:
1. Add a test class `TempInputStreamTest` for the test of `TempInputStream`
2. Add a unit test to `RqMultiPartTest` that actually uses `RqMultiPart`

Bertram25 commented:
@pinaf Thanks a lot for your review. :)

> 0.21 here

I adapted the version in the two newly created files.

> deletesTempFile()

Fixed.

> 1. Add a test class TempInputStreamTest for the test of TempInputStream

Fixed. Thanks for the hint.

> RqMultiPart can delete the underlying temporary file

Actually, as the test was indeed about `TempInputStream`, I reused your sentence as:
`TempInputStream can delete the temporary cache file when it calls close().`
in `TempInputStreamTest.java`

> 2. Add a unit test to RqMultiPartTest that actually uses RqMultiPart

The problem is I cannot know which temporary file is created as its name is random. On my system, it is something like:
`/tmp/org.takes.RqMultipart_XXXXXX.tmp` (The `XXXXXX` part is made of random digits.).
Thus, depending on the OS and the implementation of `File.createTempFile()` it is really not easily testable explaining why I created a test for the underlying body class instead to make sure the file gets deleted.
IMHO, the change is actually tested and is functional, but if you have a better idea, I'll listen to your advice.

Best regards,

pinaf commented on src/test/java/org/takes/rq/TempInputStreamTest.java:
```diff
@@ -0,0 +1,71 @@
+
+/**
+ * Test case for {@link TempInputStream}.
+ * @author Yegor Bugayenko (yegor@teamed.io)
```
@Bertram25 you should be the author here

pinaf commented on src/test/java/org/takes/rq/TempInputStreamTest.java:
```diff
@@ -0,0 +1,71 @@
+
+    /**
+     * TempInputStream can delete the temporary cache file
+     * when it calls close().
```
@Bertram25 I'd say `when closed` instead of `when it calls close()`

pinaf commented:
@Bertram25 thanks 2 comments

Bertram25 commented:
@pinaf Fixes done. Thanks for the new review. :)

pinaf commented:
@rultor merge

rultor commented:
> @rultor merge

@pinaf Thanks for your request. @yegor256 Please confirm this.

yegor256 commented on src/main/java/org/takes/rq/TempInputStream.java:
```diff
@@ -0,0 +1,98 @@
+ *
+ * <p>All implementations of this interface must be immutable and thread-safe.
+ *
+ * @author Yegor Bugayenko (yegor@teamed.io)
```
@Bertram25 I'm not the author of it :)

yegor256 commented on src/main/java/org/takes/rq/TempInputStream.java:
```diff
@@ -0,0 +1,98 @@
+    /**
+     * Original stream.
+     */
+    private final transient InputStream origin;
```
@Bertram25 brrr... you're extending `InputSteam` **and** encapsulating it? isn't inheritance good enough?

yegor256 commented:
@Bertram25 a few comments from me

Bertram25 commented:
@yegor256 Hey, thanks for the code review. :)

> @Bertram25 I'm not the author of it :)

Fixed. (At first and some time ago already, I assumed we devs were to give the code away, hence the given authorship.)

> @Bertram25 brrr... you're extending InputSteam and encapsulating it? isn't inheritance good enough?

Here, I encapsulated the origin `InputStream` to keep its functionalities in the functions overriden (read, skip, ...) (the origin `InputStream` in our specific case actually being a `CapInputStream`) and I extended `InputStream` as I simply needed it to add the desired functionalities for `TempInputStream`.

Also, note that you're doing the exact same thing in `CapInputStream.java`

If you prefer, I can have a go at extending `CapInpuStream` only instead, as it should suffice for our need. Is that the change you're requesting?

Best regards,

Bertram25 commented:
@yegor256 I'll rebase as the PR has become non mergeable.

pinaf commented:
@Bertram25 you could just merge with master manually instead of rebasing

Bertram25 commented:
@pinaf Most organizations prefer rebasing. I'll know I can just merge the next time. Thanks.

pinaf commented:
@Bertram25 what's the status here?

Bertram25 commented:
@pinaf Hi,

@yegor256 asked me why I used encapsulation and extension of the same class `InputStream`

I answered I was extending the `InputStream` class to support the file deletion handling and I encapsulated the `InputStream` *instance* as I needed the features added by the decorated object there. (The encapsulated object instance is a `CapInputStream` in our case.).
Plus, the exact same thing is done in `CapInputStream.java` (@yegor256 is the author of that file AFAIK.)

Thus, IMHO, the code is good, but if @yegor256 still would like a change, I can extend CapInputStream instead and drop the encapsulation, while I still do think the current implementation is more extensible.

So, I'm still waiting for @yegor256 's decision here.

Regards,

pinaf commented:
@Bertram25 thanks! @yegor256 please advise

pinaf commented:
@yegor256 ping

pinaf commented:
@yegor256 is this acceptable?

yegor256 commented:
@Bertram25 implementation of `CapInputStream` is also incorrect. You can easily reproduce a problem by this code:

```
InputStream stream = new CapInputStream(
  new ByteArrayInputStream(new byte[] {1, 2, 3}), 
  3
);
stream.skip(1);
assert stream.read() == 2;
```

Thus, I'm suggesting to submit a bug about `CapInputStream` and fix this branch too.

pinaf commented:
@Bertram25 ping

pinaf commented:
@Bertram25 please close the issue

pinaf commented:
@yegor256 do you mind closing this issue?

yegor256 commented:
@Bertram25 @pinaf I'm closing it, it's too old

pinaf commented:
@yegor256 why did you mark it as invalid?

yegor256 commented:
@pinaf my bad, I thought it's a ticket. removing the label now

pinaf commented:
@yegor256 nothing yet :(

pinaf commented:
@yegor256 could you please take a look at why I didn't get paid?

pinaf commented:
@yegor256 I thought CRs were paid regardless of a merge. If this policy was changed please let me know so I don't have to bug you about payment.

yegor256 commented:
@davvd pls add extra 15 mins to @pinaf

yegor256 commented:
@pinaf no changes in the policy, it's just our internal problem, sorry about it

pinaf commented:
@yegor256 it's cool. thank you.

davvd commented:
> @davvd pls add extra 15 mins to @pinaf 

@yegor256 15 mins extra was just paid to @pinaf, transaction `AP-8J079864TA4566600`
